hypothesis-python/src/hypothesis/internal/conjecture/dfa/lstar.py

```diff
@@ -195,18 +345,25 @@ def seems_right(n):
             # that allows us to distinguish the state that we ended
             # up in from the state that we should have ended up in.
 
-            suffix = string[n + 1 :]
+            l1 = self.dfa.label(states[n + 1])
+            l2 = self.dfa.label(states[n]) + string[n : n + 1]
 
-            current = dfa.label(states[n + 1])
-            actual = dfa.label(states[n]) + string[n : n + 1]
-            assert self.member(current + suffix) != self.member(actual + suffix)
-            self.__add_experiment(suffix)
+            assert self.transition(states[n], string[n]) == states[n + 1]
 
-    def __add_experiment(self, e):
-        assert e not in self.__experiment_set
-        self.experiments.append(e)
-        self.__experiment_set.add(e)
-        self.__dfa_changed()
+            ex = string[n + 1 :]
+
+            assert self.member(l1 + ex) != self.member(l2 + ex)
+
+            # Adding this experiment causes us to distinguish l1 from the state
+            # labelled with l2.
+            self.__states[states[n + 1]].experiments[ex] = self.member(l1 + ex)
+
+            # We now clear the cached details that caused us to make this error
+            # so that when we recalculate this transition we get to a
+            # (hopefully now correct) different state.
+            del self.__states[states[n]].transitions[string[n]]
+            self.__dfa_changed()
+            assert self.transition(states[n], string[n]) != states[n + 1]
```

Comment:
I'm a little concerned by the fact that this assertion invokes `transition`, which has “side effects” in the sense that it could make arbitrary decisions based on the precise arrangement of `__self_organising_states`, and then store those choices back into the transition cache.

(Right now this is mostly theoretical since you're unlikely to be updating DFAs under `-O`, but maybe it will silently become annoying after future changes?)

Reply:
This is a good point. I'll extract a variable and separate that from the assertion.